src/remote_adapter/remote_adapter_client/hmi_adapter/hmi_adapter_client.cc

```diff
@@ -0,0 +1,135 @@
+
+namespace error_codes = constants::error_codes;
+
+HmiAdapterClient::HmiAdapterClient(
+    RemoteClient *client_ptr,
+    std::vector<parameter_type> &connection_parameters, QObject *parent)
+    : QObject(parent), connection_parameters_(connection_parameters) {
+  LOG_INFO("{0}", __func__);
+
+  remote_adapter_client_ptr_ = client_ptr;
+  try {
+    future_ = exitSignal_.get_future();
+  } catch (std::future_error &e) {
+    std::cerr << __func__ << " " << e.what() << "\n" << std::flush;
+  }
+}
+
+HmiAdapterClient::~HmiAdapterClient() {
+  LOG_INFO("{0}", __func__);
+  if (listener_ptr_) {
+    try {
+      exitSignal_.set_value();
+    } catch (std::future_error &e) {
+      std::cerr << __func__ << " " << e.what() << "\n" << std::flush;
+    }
+    listener_ptr_->join();
+  }
+
+  if (isconnected_) {
+    remote_adapter_client_ptr_->content_call(constants::close_handle,
+                                             connection_parameters_);
+  }
+}
+
+void HmiAdapterClient::connect() {
+  LOG_INFO("{0}", __func__);
+  if (isconnected_ || listener_ptr_) {
+    LOG_INFO("{0} Is already connected", __func__);
+    response_type result = remote_adapter_client_ptr_->content_call(
+        constants::open_handle, connection_parameters_);
+    if (error_codes::SUCCESS == result.second)
+      emit connected();
+    return;
+  }
+
+  response_type result = remote_adapter_client_ptr_->content_call(
+      constants::open_handle, connection_parameters_);
+
+  if (error_codes::SUCCESS == result.second) {
+    try {
+      isconnected_ = true;
+      emit connected();
+      auto &future = future_;
+      listener_ptr_.reset(new std::thread([this, &future] {
+        try {
+
+          while (future.wait_for(std::chrono::milliseconds(25)) ==
+                 std::future_status::timeout) {
+            this->receive();
+          }
+
+        } catch (std::future_error &e) {
+          std::cerr << "Exception in: " << __func__ << " " << e.what() << "\n"
+                    << std::flush;
+        } catch (...) {
+          std::cerr << "Unknown Exception in: " << __func__ << "\n"
+                    << std::flush;
+        }
+      }));
+    } catch (std::exception &e) {
+      std::cerr << __func__ << " " << e.what() << "\n" << std::flush;
+    } catch (...) {
+      std::cerr << "Unknown Exception in: " << __func__ << "\n" << std::flush;
+    }
+  }
+}
+
+int HmiAdapterClient::send(const std::string &data) {
+  LOG_INFO("{0}", __func__);
+  if (isconnected_) {
+    std::vector<parameter_type> parameters(connection_parameters_);
+    parameters.push_back(std::make_pair(data, constants::param_types::STRING));
+    response_type result =
+        remote_adapter_client_ptr_->content_call(constants::send, parameters);
+    if (error_codes::SUCCESS == result.second) {
+      emit bytesWritten(data.length());
+      if (result.first.length()) {
+        QString receivedData(result.first.c_str());
+        emit textMessageReceived(receivedData);
```

Comment:
@jacobkeeler Yes, I understand that it is looking odd. 
But this piece of logic increases stability of test scripts run results by preventing fake timeout fails in scripts.

**Let me explain in details:**
As you can see separate thread was created to receive messages from SDL.
(src/remote_adapter/remote_adapter_client/hmi_adapter/hmi_adapter_client.cc:64)
But time from time SDL send a lot of messages in very short period of time.
(For example communication with HMI on start of SDL)
"Receiving thread" asks data from Remote server each 25 ms by one message.
So receiving of each portion of data from SDL by expectation system of ATF is delayed. 
It is critical.
Decreasing of mentioned delay between asks of data in "Receiving thread" 
significantly loads remote server application that leads to decrease performance of SDL.
(Remote server is running on the same host as SDL is runned)
That is why was added logic to check received data each time when ATF sends data to SDL. 
This approach make possibility to deliver data from SDL to ATF's expectation system with less delay, with less loading of Remote server.

**Example:**
Abbreviations: 
 RA - regular receiving approach with "Receiving thread" only
 R+SA - receiving approach with "Receiving thread" and "receive on send"

Let's look example of timings from a simplified case with N requests on base first 2 of them:
 1) SDL sends N requests to HMI (Example: GetCapabilities for UI, VR ...  ) 
 2) HMI receives 1-st request in 0~25 ms (for both (RA and R+SA) approaches)
 3) HMI responds on 1-st request in x ms and receives next request from SDL if R+SA applied (see next line of sequence) 
 4) HMI receives 2-nd request in 25 ms (RA) or in min(25, x) ms (R+SA)
 ...
If even x > 25 it is triggered receive procedure for 3-rd request 
which starts to processing earlier and ATF send new request or response faster and so on. 
That significantly speed up receiving of messages from SDL by ATF.

As you can note R+SA become more efficient on massive communications SDL <-> HMI. 
Also this approach leads:
 - to speed up SDL <-> HMI communications without permanent loading of Remote server
 - to avoiding of fake fails of test scripts with timeout reason

 Sorry, for so long answer.
I tried to explain the decision as clear as it is possible.

Reply:
@aderiabin Couldn't this be fixed by reading all pending messages before waiting again? From my understanding, you are saying that this adapter only reads one message from the connection every 25ms without this code, but it seems to me that the wait shouldn't be necessary unless there is no more data to read.

Reply:
Yes, you are right this adapter reads one message from the connection every 25ms without mentioned code.
In another words adapter asks data from remote server each 25ms using its method `HmiAdapterClient::receive()` which uses `RemoteClient::content_call()` that return only one message on call. So as I can see it is exists two approaches to receive all pending messages from server without delay:
 - to  implement not trivial chain call logic of RemoteClient::content_call() in  HmiAdapterClient::receive() method 
 - to create new method in RemoteClient  that be able to return collection of messages and its support on server side (significant change)
Both of them are required significant amount of testing activities (run all tests from https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2304)

Could we merge existing tested approach with receiving message on send now?
After that we prepare and test one of described  above approaches and deliver it as improvement?

Reply:
@aderiabin I do see that this will be a bit more complicated than I originally thought, so I would agree that this could be implemented as an improvement at a later time.